File: twidere/src/main/java/org/mariotaku/twidere/loader/ParcelableUserLoader.java
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.loader;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.os.Bundle;
import android.support.v4.content.AsyncTaskLoader;
import android.text.TextUtils;
import android.util.Log;
import android.util.Pair;

import org.mariotaku.abstask.library.TaskStarter;
import org.mariotaku.sqliteqb.library.Columns;
import org.mariotaku.sqliteqb.library.Expression;
import org.mariotaku.twidere.Constants;
import org.mariotaku.microblog.library.MicroBlog;
import org.mariotaku.microblog.library.MicroBlogException;
import org.mariotaku.microblog.library.twitter.model.User;
import org.mariotaku.twidere.fragment.UserFragment;
import org.mariotaku.twidere.model.ParcelableAccount;
import org.mariotaku.twidere.model.ParcelableCredentials;
import org.mariotaku.twidere.model.ParcelableUser;
import org.mariotaku.twidere.model.ParcelableUserCursorIndices;
import org.mariotaku.twidere.model.ParcelableUserValuesCreator;
import org.mariotaku.twidere.model.SingleResponse;
import org.mariotaku.twidere.model.UserKey;
import org.mariotaku.twidere.model.util.ParcelableCredentialsUtils;
import org.mariotaku.twidere.model.util.ParcelableUserUtils;
import org.mariotaku.twidere.model.util.UserKeyUtils;
import org.mariotaku.twidere.provider.TwidereDataStore.CachedUsers;
import org.mariotaku.twidere.task.UpdateAccountInfoTask;
import org.mariotaku.twidere.util.MicroBlogAPIFactory;
import org.mariotaku.twidere.util.TwitterWrapper;
import org.mariotaku.twidere.util.UserColorNameManager;
import org.mariotaku.twidere.util.dagger.GeneralComponentHelper;

import javax.inject.Inject;

import static org.mariotaku.twidere.util.ContentValuesCreator.createCachedUser;

public final class ParcelableUserLoader extends AsyncTaskLoader<SingleResponse<ParcelableUser>> implements Constants {

    private final boolean mOmitIntentExtra, mLoadFromCache;
    private final Bundle mExtras;
    private final UserKey mAccountKey;
    private final UserKey mUserKey;
    private final String mScreenName;

    @Inject
    UserColorNameManager mUserColorNameManager;

    public ParcelableUserLoader(final Context context, final UserKey accountKey,
                                final UserKey userKey, final String screenName,
                                final Bundle extras, final boolean omitIntentExtra,
                                final boolean loadFromCache) {
        super(context);
        GeneralComponentHelper.build(context).inject(this);
        this.mOmitIntentExtra = omitIntentExtra;
        this.mLoadFromCache = loadFromCache;
        this.mExtras = extras;
        this.mAccountKey = accountKey;
        this.mUserKey = userKey;
        this.mScreenName = screenName;
    }

    @Override
    public SingleResponse<ParcelableUser> loadInBackground() {
        final Context context = getContext();
        final ContentResolver resolver = context.getContentResolver();
        final UserKey accountKey = mAccountKey;
        ParcelableCredentials credentials = null;
        for (ParcelableCredentials cred : ParcelableCredentialsUtils.getCredentialses(context)) {
            if (cred.account_key.equals(accountKey)) {
                credentials = cred;
                break;
            } else if (cred.account_user != null && cred.account_user.account_key.equals(accountKey)) {
                credentials = cred;
                break;
            }
        }
        if (credentials == null) return SingleResponse.getInstance();
        if (!mOmitIntentExtra && mExtras != null) {
            final ParcelableUser user = mExtras.getParcelable(EXTRA_USER);
            if (user != null) {
                final ContentValues values = ParcelableUserValuesCreator.create(user);
                resolver.insert(CachedUsers.CONTENT_URI, values);
                ParcelableUserUtils.updateExtraInformation(user, credentials, mUserColorNameManager);
                return SingleResponse.getInstance(user);
            }
        }
        final MicroBlog twitter = MicroBlogAPIFactory.getTwitterInstance(context, credentials, true, true);
        if (twitter == null) return SingleResponse.getInstance();
        if (mLoadFromCache) {
            final Expression where;
            final String[] whereArgs;
            if (mUserKey != null) {
                where = Expression.equalsArgs(CachedUsers.USER_KEY);
                whereArgs = new String[]{mUserKey.toString()};
            } else if (mScreenName != null) {
                final String host = mAccountKey.getHost();
                if (host != null) {
                    where = Expression.and(
                            Expression.likeRaw(new Columns.Column(CachedUsers.USER_KEY), "'%@'||?"),
                            Expression.equalsArgs(CachedUsers.SCREEN_NAME)
                    );
                    whereArgs = new String[]{host, mScreenName};
                } else {
                    where = Expression.equalsArgs(CachedUsers.SCREEN_NAME);
                    whereArgs = new String[]{mScreenName};
                }
            } else {
                return SingleResponse.getInstance();
            }
            final Cursor cur = resolver.query(CachedUsers.CONTENT_URI, CachedUsers.COLUMNS,
                    where.getSQL(), whereArgs, null);
            if (cur != null) {
                try {
                    cur.moveToFirst();
                    final ParcelableUserCursorIndices indices = new ParcelableUserCursorIndices(cur);
                    while (!cur.isAfterLast()) {
                        final ParcelableUser user = indices.newObject(cur);
                        if (TextUtils.equals(UserKeyUtils.getUserHost(user), user.key.getHost())) {
                            user.account_key = accountKey;
                            user.account_color = credentials.color;
                            return SingleResponse.getInstance(user);
                        }
                        cur.moveToNext();
                    }
                } finally {
                    cur.close();
                }
            }
        }
        try {
            final User twitterUser;
            if (mExtras != null && UserFragment.Referral.SELF_PROFILE.equals(mExtras.getString(EXTRA_REFERRAL))) {
                twitterUser = twitter.verifyCredentials();
            } else {
                String profileUrl = null;
                if (mExtras != null) {
                    profileUrl = mExtras.getString(EXTRA_PROFILE_URL);
                }
                if (MicroBlogAPIFactory.isStatusNetCredentials(credentials) && mUserKey != null &&
                        profileUrl != null && !TextUtils.equals(credentials.account_key.getHost(),
                        mUserKey.getHost())) {
                    twitterUser = twitter.showExternalProfile(profileUrl);
                } else {
                    final String id = mUserKey != null ? mUserKey.getId() : null;
                    twitterUser = TwitterWrapper.tryShowUser(twitter, id, mScreenName,
                            credentials.account_type);
                }
            }
            final ContentValues cachedUserValues = createCachedUser(twitterUser);
            resolver.insert(CachedUsers.CONTENT_URI, cachedUserValues);
            final ParcelableUser user = ParcelableUserUtils.fromUser(twitterUser, accountKey);
            ParcelableUserUtils.updateExtraInformation(user, credentials, mUserColorNameManager);
            final SingleResponse<ParcelableUser> response = SingleResponse.getInstance(user);
            response.getExtras().putParcelable(EXTRA_ACCOUNT, credentials);
            return response;
        } catch (final MicroBlogException e) {
            Log.w(LOGTAG, e);
            return SingleResponse.getInstance(e);
        }
    }

    @Override
    protected void onStartLoading() {
        if (!mOmitIntentExtra && mExtras != null) {
            final ParcelableUser user = mExtras.getParcelable(EXTRA_USER);
            if (user != null) {
                deliverResult(SingleResponse.getInstance(user));
            }
        }
        forceLoad();
    }

    @Override
    public void deliverResult(SingleResponse<ParcelableUser> data) {
        super.deliverResult(data);
        if (data.hasData()) {
            final ParcelableUser user = data.getData();
            if (user.is_cache) return;
            final ParcelableAccount account = data.getExtras().getParcelable(EXTRA_ACCOUNT);
            if (account != null) {
                final UpdateAccountInfoTask task = new UpdateAccountInfoTask(getContext());
                task.setParams(Pair.create(account, user));
                TaskStarter.execute(task);
            }
        }
    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/loader/ParcelableUserLoader.kt
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.loader

import android.accounts.AccountManager
import android.content.Context
import android.os.Bundle
import androidx.loader.content.FixedAsyncTaskLoader
import android.text.TextUtils
import android.util.Log
import org.mariotaku.abstask.library.TaskStarter
import org.mariotaku.ktextension.set
import org.mariotaku.library.objectcursor.ObjectCursor
import org.mariotaku.microblog.library.MicroBlog
import org.mariotaku.microblog.library.MicroBlogException
import org.mariotaku.microblog.library.mastodon.Mastodon
import org.mariotaku.microblog.library.twitter.model.Paging
import org.mariotaku.sqliteqb.library.Columns
import org.mariotaku.sqliteqb.library.Expression
import org.mariotaku.twidere.R
import org.mariotaku.twidere.TwidereConstants.*
import org.mariotaku.twidere.annotation.AccountType
import org.mariotaku.twidere.extension.api.tryShowUser
import org.mariotaku.twidere.extension.model.api.mastodon.toParcelable
import org.mariotaku.twidere.extension.model.api.toParcelable
import org.mariotaku.twidere.extension.model.host
import org.mariotaku.twidere.extension.model.isAcctPlaceholder
import org.mariotaku.twidere.extension.model.newMicroBlogInstance
import org.mariotaku.twidere.model.AccountDetails
import org.mariotaku.twidere.model.ParcelableUser
import org.mariotaku.twidere.model.SingleResponse
import org.mariotaku.twidere.model.UserKey
import org.mariotaku.twidere.model.util.AccountUtils
import org.mariotaku.twidere.model.util.ParcelableUserUtils
import org.mariotaku.twidere.provider.TwidereDataStore.CachedUsers
import org.mariotaku.twidere.task.UpdateAccountInfoTask
import org.mariotaku.twidere.util.UserColorNameManager
import org.mariotaku.twidere.util.dagger.GeneralComponent
import javax.inject.Inject

class ParcelableUserLoader(
        context: Context,
        private val accountKey: UserKey?,
        private val userKey: UserKey?,
        private val screenName: String?,
        private val extras: Bundle?,
        private val omitIntentExtra: Boolean,
        private val loadFromCache: Boolean
) : FixedAsyncTaskLoader<SingleResponse<ParcelableUser>>(context) {

    private val profileImageSize = context.getString(R.string.profile_image_size)

    @Inject
    lateinit var userColorNameManager: UserColorNameManager

    init {
        GeneralComponent.get(context).inject(this)
    }

    override fun loadInBackground(): SingleResponse<ParcelableUser> {
        val context = context
        val resolver = context.contentResolver
        val accountKey = accountKey ?: return SingleResponse(MicroBlogException("No account"))
        val am = AccountManager.get(context)
        val details = AccountUtils.getAllAccountDetails(am, AccountUtils.getAccounts(am), true).firstOrNull {
            if (it.key == accountKey) {
                return@firstOrNull true
            } else if (it.user.account_key == accountKey) {
                return@firstOrNull true
            }
            return@firstOrNull false
        } ?: return SingleResponse()
        if (!omitIntentExtra && extras != null) {
            val user = extras.getParcelable<ParcelableUser?>(EXTRA_USER)
            if (user != null) {
                val values = ObjectCursor.valuesCreatorFrom(ParcelableUser::class.java).create(user)
                resolver.insert(CachedUsers.CONTENT_URI, values)
                ParcelableUserUtils.updateExtraInformation(user, details, userColorNameManager)
                return SingleResponse(user).apply {
                    extras[EXTRA_ACCOUNT] = details
                }
            }
        }
        if (loadFromCache) {
            val where: Expression
            val whereArgs: Array<String>
            if (userKey != null) {
                where = Expression.equalsArgs(CachedUsers.USER_KEY)
                whereArgs = arrayOf(userKey.toString())
            } else if (screenName != null) {
                val host = accountKey.host
                if (host != null) {
                    where = Expression.and(
                            Expression.likeRaw(Columns.Column(CachedUsers.USER_KEY), "'%@'||?"),
                            Expression.equalsArgs(CachedUsers.SCREEN_NAME))
                    whereArgs = arrayOf(host, screenName)
                } else {
                    where = Expression.equalsArgs(CachedUsers.SCREEN_NAME)
                    whereArgs = arrayOf(screenName)
                }
            } else {
                return SingleResponse()
            }
            resolver.query(CachedUsers.CONTENT_URI, CachedUsers.COLUMNS, where.sql,
                    whereArgs, null)?.let { cur ->
                @Suppress("ConvertTryFinallyToUseCall")
                try {
                    cur.moveToFirst()
                    val indices = ObjectCursor.indicesFrom(cur, ParcelableUser::class.java)
                    while (!cur.isAfterLast) {
                        val user = indices.newObject(cur)
                        if (TextUtils.equals(user.host, user.key.host)) {
                            user.account_key = accountKey
                            user.account_color = details.color
                            return SingleResponse(user).apply {
                                extras[EXTRA_ACCOUNT] = details
                            }
                        }
                        cur.moveToNext()
                    }
                } finally {
                    cur.close()
                }
            }
        }
        try {
            val user = when (details.type) {
                AccountType.MASTODON -> showMastodonUser(details)
                else -> showMicroBlogUser(details)
            }
            val creator = ObjectCursor.valuesCreatorFrom(ParcelableUser::class.java)
            val cachedUserValues = creator.create(user)
            resolver.insert(CachedUsers.CONTENT_URI, cachedUserValues)
            ParcelableUserUtils.updateExtraInformation(user, details, userColorNameManager)
            return SingleResponse(user).apply {
                extras[EXTRA_ACCOUNT] = details
            }
        } catch (e: MicroBlogException) {
            Log.w(LOGTAG, e)
            return SingleResponse(exception = e)
        }

    }

    private fun showMastodonUser(details: AccountDetails): ParcelableUser {
        val mastodon = details.newMicroBlogInstance(context, Mastodon::class.java)
        if (userKey == null) throw MicroBlogException("Invalid user id")
        if (!userKey.isAcctPlaceholder) {
            return mastodon.getAccount(userKey.id).toParcelable(details)
        }
        if (screenName == null) throw MicroBlogException("Screen name required")
        val resultItem = mastodon.searchAccounts("$screenName@${userKey.host}", Paging().count(1))
                .firstOrNull() ?: throw MicroBlogException("User not found")
        return resultItem.toParcelable(details)
    }

    private fun showMicroBlogUser(details: AccountDetails): ParcelableUser {
        val microBlog = details.newMicroBlogInstance(context, MicroBlog::class.java)
        val profileUrl = extras?.getString(EXTRA_PROFILE_URL)
        val response = if (extras != null && extras.getBoolean(EXTRA_IS_ACCOUNT_PROFILE)) {
            microBlog.verifyCredentials()
        } else if (details.type == AccountType.STATUSNET && userKey != null && profileUrl != null
                && details.key.host != userKey.host) {
            microBlog.showExternalProfile(profileUrl)
        } else {
            microBlog.tryShowUser(userKey?.id, screenName, details.type)
        }
        return response.toParcelable(details, profileImageSize = profileImageSize)
    }

    override fun onStartLoading() {
        if (!omitIntentExtra && extras != null) {
            val user = extras.getParcelable<ParcelableUser>(EXTRA_USER)
            if (user != null) {
//                deliverResult(SingleResponse(user))
            }
        }
        forceLoad()
    }

    override fun deliverResult(data: SingleResponse<ParcelableUser>?) {
        super.deliverResult(data)
        val user = data?.data ?: return
        if (user.is_cache) return
        val account = data.extras.getParcelable<AccountDetails>(EXTRA_ACCOUNT)
        if (account != null) {
            val task = UpdateAccountInfoTask(context)
            task.params = Pair(account, user)
            TaskStarter.execute(task)
        }
    }
}
